### ISSUE TITLE
most key requests fail to work or behave as if Home key was sent

### ISSUE THREAD

**Issue by johngo7470** [Toxblh/samsung-tv-control#9]
I installed samsung-tv-control, and picked up the token after the first execution, and updated my script with the token.

I updated the same script by removing the requests for getAppsFromTV and openApp.  

When I run the script, I get what appears to be expected output from the script, but the tv usually behaves as if I had sent KEYS.KEY_HOME. 

For example, when my script is:

```
const Samsung = require('samsung-tv-control').default
const { KEYS } = require('samsung-tv-control/lib/keys')
const { APPS } = require('samsung-tv-control/lib/apps')

const config = {
  debug: true, // Default: false
  ip: '192.168.1.162',
  mac: '84C0EF0BAC2E',
  name: 'NodeJS-Test', // Default: NodeJS
  port: 8002, // Default: 8002
  token: '56918168',
}

const control = new Samsung(config)

control.turnOn()
control
  .isAvaliable()
  .then(() => {
    // Get token for API
    control.getToken(token => {
      console.info('# Response getToken:', token)
    })

    // Send key to TV
    control.sendKey(KEYS.KEY_POWEROFF, function(err, res) {
      if (!err) {
        console.log('# Response sendKey', res)
      }
    })

  })
  .catch(e => console.error(e))
```

I receive the following output:

```
pi@raspi:~/bin $ node samsungCTL
TV is avaliable
TV is avaliable
message {
  "data": {
    "clients": [
      {
        "attributes": {
          "name": "Tm9kZUpTIFJlbW90ZQ== ",
          "token": "56918168"
        },
        "connectTime": 1577047409542,
        "deviceName": "Tm9kZUpTIFJlbW90ZQ== ",
        "id": "1cc0fc1e-855b-489d-b879-49c5614d4765",
        "isHost": false
      }
    ],
    "id": "1cc0fc1e-855b-489d-b879-49c5614d4765"
  },
  "event": "ms.channel.connect"
}
# Response getToken: null
message {
  "data": {
    "attributes": {
      "name": "Tm9kZUpTIFJlbW90ZQ== ",
      "token": "56918168"
    },
    "connectTime": 1577047409581,
    "deviceName": "Tm9kZUpTIFJlbW90ZQ== ",
    "id": "4e64cf30-7723-47a-bd57-8c3d3a725b28",
    "isHost": false
  },
  "event": "ms.channel.clientConnect"
}
message {
  "data": {
    "clients": [
      {
        "attributes": {
          "name": "Tm9kZUpTIFJlbW90ZQ== ",
          "token": "56918168"
        },
        "connectTime": 1577047409542,
        "deviceName": "Tm9kZUpTIFJlbW90ZQ== ",
        "id": "1cc0fc1e-855b-489d-b879-49c5614d4765",
        "isHost": false
      },
      {
        "attributes": {
          "name": "Tm9kZUpTIFJlbW90ZQ== ",
          "token": "56918168"
        },
        "connectTime": 1577047409581,
        "deviceName": "Tm9kZUpTIFJlbW90ZQ== ",
        "id": "4e64cf30-7723-47a-bd57-8c3d3a725b28",
        "isHost": false
      }
    ],
    "id": "4e64cf30-7723-47a-bd57-8c3d3a725b28"
  },
  "event": "ms.channel.connect"
}
# Response sendKey { data:
   { clients: [ [Object], [Object] ],
     id: '4e64cf30-7723-47a-bd57-8c3d3a725b28' },
  event: 'ms.channel.connect' }
message {
  "data": {
    "attributes": {
      "name": "Tm9kZUpTIFJlbW90ZQ== ",
      "token": "56918168"
    },
    "connectTime": 1577047409542,
    "deviceName": "Tm9kZUpTIFJlbW90ZQ== ",
    "id": "1cc0fc1e-855b-489d-b879-49c5614d4765",
    "isHost": false
  },
  "event": "ms.channel.clientDisconnect"
}
```

Should I see the command I sent, in the output? (or does the tv simply accept the command without confirmation/echo?)

If there is no indication of an error, why are my commands mostly failing, or resulting in behavior on the tv as if I had sent the KEYS.KEY_HOME command? 

Does there need to be a delay between the getToken command and any sendKey commands?

Now that I know the token, is the getToken necessary in the script?

I tried using multiple sendKeys in the script:

```
control.turnOn()
control
  .isAvaliable()
  .then(() => {
    // Get token for API
    control.getToken(token => {
      console.info('# Response getToken:', token)
    })

    // Send key to TV
    control.sendKey(KEYS.KEY_VOLDOWN, function(err, res) {
      if (!err) {
        console.log('# Response sendKey', res)
      }
    })
    control.sendKey(KEYS.KEY_VOLDOWN, function(err, res) {
      if (!err) {
        console.log('# Response sendKey', res)
      }
    })
    control.sendKey(KEYS.KEY_VOLDOWN, function(err, res) {
      if (!err) {
        console.log('# Response sendKey', res)
      }
    })
    control.sendKey(KEYS.KEY_VOLDOWN, function(err, res) {
      if (!err) {
        console.log('# Response sendKey', res)
      }
    })
    control.sendKey(KEYS.KEY_VOLDOWN, function(err, res) {
      if (!err) {
        console.log('# Response sendKey', res)
      }
    })
    control.sendKey(KEYS.KEY_VOLDOWN, function(err, res) {
      if (!err) {
        console.log('# Response sendKey', res)
      }
    })
    control.sendKey(KEYS.KEY_VOLDOWN, function(err, res) {
      if (!err) {
        console.log('# Response sendKey', res)
      }
    })
    control.sendKey(KEYS.KEY_VOLDOWN, function(err, res) {
      if (!err) {
        console.log('# Response sendKey', res)
      }
    })

  })
  .catch(e => console.error(e))
```
The TV only seemed to execute one tick of VolumeDown, and the output from the script was, in a word, peculiar. I'm wondering if I'm trying to send multiple commands in the wrong way. 

Also, I did a wireshark capture, and noticed some packets going to 8001 and some going to 8002 - is this normal?

Incidentally, the home page example script is updated with const Samsung = require('samsung-tv-control')**_.default_**, but the sample index.js in the example folder, still has: const Samsung = require('samsung-tv-control')  (I saw in one of the Issues, that **.default** is missing from the example script in the examples folder)

**Comment by johngo7470** [comment]
With more testing, I discovered answers to some of my questions...

The KEYS.KEY_HOME I was seeing, was actually within getToken(), which I no longer need. 

The Key that I WAS sending, was seemingly not doing much, except triggering a client.disconnect.

When I re-wrote the script as:

```
control.turnOn()
control
  .isAvaliable()
  .then(() => {
    control.sendKey(KEYS.KEY_VOLUP, function(err, res) {
      if (!err) {
        console.log('# Response sendKey', res)
      }
    })
    control.sendKey(KEYS.KEY_VOLUP, function(err, res) {
      if (!err) {
        console.log('# Response sendKey', res)
      }
    })
  })
```

does send a VOLUP, though when sending one VOLUP, the wss session stays open.  If I send a second command of any kind (like in the example code above), the wss session gets closed (the script completes and returns to a prompt), and only one command appears to be executed (presumably the first VOLUP). 

Part of the problem earlier, was lack of familiarity of javascript (a.k.a. "loose nut behind the wheel").

I'm still unsure how to send multiple commands (e.g. 20 VOLUP commands, to set a medium volume level), and how to end a wss session, or if I can leave the session open to save time, connecting.

If I figure out the error of my ways, I'll post details here...

**Comment by increddibelly** [comment]
remember VOLUP is relative, so if it's already loud, you'll be making it 20 steps louder...

**Comment by Toxblh** [comment]
Ok, Ill check this today after 8pm GMT

**Comment by Toxblh** [comment]
So. Step-by-step.
1. Yes, you need to get a token for the first time. Just for save it in config
2. Traffic to 8001 port it is checking availability tv, just HTTP request to debug page
3. Between commands need to add delays. around 100-200ms.
4. Fix 2 step and do just KEY_VOLUP 20 times. Will be it a new version together with support models 2015 and older. Will be more methods include all async/await + will be a queue of commands.

**Comment by johngo7470** [comment]
Thank you for the clarifications.

Good to know (delay between commands) - thank you!

As for #4... sorry, should have mentioned I have a 2017. 

I think I can figure out how to send multiple commands by studying the code a bit more, and also by studying a similar python script. 

One thing I think that's confusing me... if you could clarify... currently, every time _sent is called in index.js, it creates a new ws (websocket) object, which isn't the direction I want to go, to send multiple key presses.

In theory, I probably want to open a websocket, keep the object, and as long as it's in a readyState of 1, just send the key press, else open a new socket, if the old one is no longer viable, yes?

**Comment by Toxblh** [comment]
@johngo7470  about keep 1 socket open for sending all commands, it makes sense, and that in my roadmap for future. 

So for good exp you can use the code now

```
async function main() {
  await control.sendKeyPromise(KEYS.KEY_1)
  await control.sendKeyPromise(KEYS.KEY_3)
  await control.sendKeyPromise(KEYS.KEY_2)
  await control.sendKeyPromise(KEYS.KEY_ENTER)
}

main()
```

**Comment by johngo7470** [comment]
Agreed, a good way to send multiple keys!

I heavily console.log'd index.js, to reveal what was going on, and now understand that multiple websockets were being opened, using your sample code.

In the end, admittedly, I switched to a python solution, because I found that the startup time for Node, was quite high.  I was unable to get APE's solution working, but between his code and your solution, I was able to understand enough to write very small scripts to perform the functions I needed.

Many thanks for your code - it helped explain a great many things, and helped me to write exactly what I needed!
